src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs

```diff
@@ -667,42 +703,40 @@ protected StatementSyntax GetStatementContainingInvocationToExtractedMethodWorke
-                var methodOperation = semanticModel.GetOperation(methodDeclaration, cancellationToken);
 
-                var returnOperations = methodOperation.DescendantsAndSelf().OfType<IReturnOperation>();
-
-                foreach (var returnOperation in returnOperations)
+                if (syntaxNode is MethodDeclarationSyntax methodDeclaration)
                 {
-                    // If thereturn statement is located in a nested local function or lambda it
-                    // shouldn't contribute to the nullability of the extracted method's return type
-                    if (!ReturnOperationBelongsToMethod(returnOperation.Syntax, methodOperation.Syntax))
+                    var nullableReturnOperations = await CheckReturnOperations(methodDeclaration).ConfigureAwait(false);
+                    if (nullableReturnOperations is object)
                     {
-                        continue;
+                        return nullableReturnOperations;
                     }
 
-                    var syntax = returnOperation.ReturnedValue?.Syntax ?? returnOperation.Syntax;
-                    var returnTypeInfo = semanticModel.GetTypeInfo(syntax, cancellationToken);
-                    if (returnTypeInfo.Nullability.FlowState == NullableFlowState.MaybeNull)
+                    var newDocument = await GenerateNewDocument(methodDeclaration.ReturnType).ConfigureAwait(false);
+                    return await SemanticDocument.CreateAsync(newDocument, cancellationToken).ConfigureAwait(false);
+                }
+                else if (syntaxNode is LocalFunctionStatementSyntax localMethodDeclaration)
+                {
+                    var nullableReturnOperations = await CheckReturnOperations(localMethodDeclaration).ConfigureAwait(false);
+                    if (nullableReturnOperations != null)
                     {
-                        // Flow state shows that return is correctly nullable
-                        return await base.UpdateMethodAfterGenerationAsync(originalDocument, methodSymbolResult, cancellationToken).ConfigureAwait(false);
+                        return nullableReturnOperations;
                     }
-                }
-
-                // Return type can be updated to not be null
-                var newType = methodSymbolResult.Data.ReturnType.WithNullability(NullableAnnotation.NotAnnotated);
 
-                var oldRoot = await originalDocument.Document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-                var newRoot = oldRoot.ReplaceNode(methodDeclaration.ReturnType, newType.GenerateTypeSyntax());
-
-                var newDocument = originalDocument.Document.WithSyntaxRoot(newRoot);
-                return await SemanticDocument.CreateAsync(newDocument, cancellationToken).ConfigureAwait(false);
+                    var newDocument = await GenerateNewDocument(localMethodDeclaration.ReturnType).ConfigureAwait(false);
```

Comment:
this loic is also identical.  the only difference is in which returnType node you use.  So this can just be:

```
var returnType = syntaxNode is MethodDeclaration method ? method.ReturnType : ((LocalFunction)syntaxNode).ReturnType;
```

That's the only difference.  Everything else can be shared/identical.